added co-effects testing

Cyrik opened the pull request:
- added some tests for sanity
- added two ways to do resolve caching
- added co-effects

Addresses parts of #3

Commit:
added co-effects testing

Cyrik commented:
I've played around with https://github.com/metosin/sieppari but unless we want to do more with it than just co-effects it would just add a lot of complications. All of the co-effects would live in a single interceptor anyway.

I've played with some different designs, but I like this one the best. The namespaced keywords work well in IDEs, there are no name clashes, there is no reason to define the new "input key", finding the implementation of the co-effect is easy.

jcpsantiago commented on example/src/example/main.clj:
```diff
@@ -93,7 +93,7 @@
   (def res
     (-> req
-        b/preprocessed)))
+        (b/preprocessed 'example.main))))
```
I'll fix this in the main branch anyways thanks for reminding me 😅

jcpsantiago commented on test/jcpsantiago/bulgogi_test.clj:
```diff
@@ -1,9 +1,90 @@
   (:require
     [clojure.test :refer :all]
-    [jcpsantiago.bulgogi :refer :all]))
+    [jcpsantiago.bulgogi :as SUT]
```
just curious how did you come up with SUT?

Cyrik replied:
I've been using it for a few years, [SUT](https://en.wikipedia.org/wiki/System_under_test).
I like being able to immediately spot which parts of the test are the thing we are actually testing. Having the same naming helps and the upper case makes it easy to spot.

jcpsantiago replied:
cool, I wasn't aware of this 👍 sounds like a good concept indeed

jcpsantiago commented on test/jcpsantiago/bulgogi_test.clj:
```diff
@@ -1,9 +1,90 @@
+       (some #(= 1 %))
+       boolean->int))
+
+(defn ^{:bulgogi/co-effect ::added} needs-co-effect
```
this looks good 👍 I also like the fact it's using namespaced keywords, makes it clear, explicit and self-documenting; the interface is still super simple, so anyone can learn it by glancing an example without much fuss, which is perfect :)

jcpsantiago replied:
I wonder if it should just be called a `dep` though, because we're not using interceptors at all? This language might be alien to people unfamiliar with interceptors/re-frame

Cyrik replied:
It kind of depends, since coeffect is not a re-frame specific naming scheme. It's based on coeffect context theory. There's a great explanation here: http://tomasp.net/coeffects/
We should name it `jcpsantiago.bulgogi/coeffect` though. But if you think dependency is easier I'm not going to die on this hill :)

jcpsantiago replied:
that's such a cool website! I didn't read a theoretical explanation about coeffects before, thanks.
I like the name `dependency`, but at the same time `coeffect` is less handholding and people can also read the docs lol let's go with your proposal `jcpsantiago.bulgogi/coeffect`

Commit:
Merge branch 'main' into pr/4

jcpsantiago committed:
Merge branch 'main' into wip/co-effects

jcpsantiago commented on src/jcpsantiago/bulgogi.clj:
```diff
@@ -31,7 +48,8 @@
+        co-effects (memoized-co-effects fns)
         fn-ks (map keyword features)]
-    (->> (transformed input-data fns)
+    (->> (transformed (enriched input-data co-effects) fns)
```
we are limited to one dependency level, right? e.g. say `a` needs `b` which needs `c`, `enriched` will try to apply `b` and `c` to the data, but it's possible they are not executed in the correct order (if I'm seeing this correctly). There's no problem for simple relationships `a<-b`

Cyrik replied:
Yes, I've built it to be one level right now. Recursing on the dependencies would be easy to implement though.
Currently coeffect functions are treated differently than feature functions, they don't get to have rely on coeffects themselves. I wasn't sure how complicated the system is allowed to be, so I started with the minimum feature set.
I'd be happy to extend it so that coeffect functions can rely on other coeffects, the basic version is pretty trivial to implement. There are some design decisions about circular dependencies and idempotence though.

jcpsantiago replied:
yeah I think this is a great start, I'm also not sure how complex this needs to be in the real world.. from my experience at Billie this is already enough

Cyrik commented:
Ah one more thing I was thinking about:
`^{:bulgogi/co-effect ::added}` could also be `^{:bulgogi/co-effect (inject-coeffect ::added params)}` so that coeffect functions can be parameterized. This way if you have a `::call-api-service` you can pass it an id or something and don't need to create as many handlers.
Not sure if this is too complicated for the users of the system.

jcpsantiago commented on test/jcpsantiago/bulgogi_test.clj:
```diff
@@ -1,9 +1,90 @@
+  data)
+
+(defn added [_]
+  {::added "some data"})
```
also just realised this solution would make it mandatory that functions always output a map with a namespaced keyword, right? I think it's a good idea, makes all functions have "the same type"

Cyrik replied:
Nope. Coeffect functions have to return a map with a namespaced keyword because their result gets merged into the input-data map. As a convention, the keyword should match their name, so that anyone relying on the coeffect already knows where the data is found.

```clojure
(defn ^{:bulgogi/co-effect ::added} needs-co-effect
  [{data ::added}] ;; the ::added thing here comes from the ::added coeffect
    data)

(defn added [_]
  {::added "some data"}) ;; this gets merged into the :input-data map
```

Feature functions can still return whatever they want.

jcpsantiago replied:
I tried making a pair myself, works well. The only thing I'm not a super fan of is the fact that, if there is a different signature between a "feature" and a "coeffect", a data scientist using Bulgogi would need to know that that's how the system works and figure out if they are building a feature or a coeffect (I agree this should be obvious in most cases), additionally it's possible for this to evolve over time and feature become coeffects. 

I'm just speculating here :D right now there are 0 users for Bulgogi, so I think it's ok to assume a hypothetical user would know this is how it works.

Cyrik replied:
I think you might be right here. In re-frame it's not as confusing because you add one by calling `reg-cofx` and the final handler with `reg-event-fx`. Bulgogi just registers them magically so it's harder to know what you are building.

Conceptually a coeffect is supposed to enrich the data with some extra data, potentially with extra sideeffects. Feature handlers are supposed to just return a calculated feature of the input. But I see how it could be confusing.

I'll take some hammock time to think about just combining them into the same thing. The only negative that I see immediately is that we would have to solve the dependency graph problem from above, but that's not the end of the world.

Commit:
redo normal feature as feature<>coeffect pair

Commit:
rename co-effect to coeffect

Cyrik committed:
Merge branch 'jcpsantiago:main' into wip/co-effects

jcpsantiago commented:
Moved it out of draft, because I like the state of the solution. I would like to expand the `/example` with this and find more problems/awkwardnesses.

Commit:
added auto-loading

jcpsantiago commented on src/jcpsantiago/bulgogi.clj:
```diff
@@ -31,7 +55,27 @@
+       (mapcat keys)
+       distinct?))
+
+(all-features)
```
was this just for testing, or is there another behind calling it here?

Cyrik replied:
This is my current try for compile time collision checking. Requiring the ns should throw a compile time error if you have a collision

Cyrik commented:
I'll add some descriptions + coeffect auto-loading + using the auto-load tomorrow. Just wanted to push it so you can have a look at what I'm thinking.
I think with the auto-load it makes sense to keep coeffects and features as two separate things for now. We should probably merge into master tomorrow since this wouldn't break anything, just add features.

Then a new branch can start to play around with multi-layer coeffects.

@jcpsantiago What do you think?

Commit:
auto-load coeffects

Cyrik commented:
So auto-loading works for both now.
Some of the code is still a bit ugly, but I think this can be pulled in so we can play around with branches based on this.

Commit:
remove empty comment

jcpsantiago reviewed (changes requested):
small comment left before merging

Cyrik commented on deps.edn:
```diff
@@ -1,5 +1,5 @@
 {:paths ["src" "resources"]
- :deps {org.clojure/clojure {:mvn/version "1.10.1"}}
+ :deps {org.clojure/clojure {:mvn/version "1.11.0-alpha4"}}
```
I'm using update-k to return namespaced keywords for the coeffect map, but we can just use reduce-kv instead if you want

jcpsantiago replied:
I would prefer `reduce-kv` for now then. I see it in the code now, it makes sense. We can revisit this later since I feel like that function will be refactored anyways.

Cyrik replied:
sure, done. And you are right, it's pretty bad, I just didn't want to clean to much before we don't know where we are going :)

Cyrik committed:
downgrade to 1.10.1

Cyrik committed:
Merge branch 'wip/co-effects' of https://github.com/Cyrik/bulgogi into wip/co-effects

jcpsantiago reviewed (approved):
Thank you for a great first PR and great discussions @Cyrik ! 🚀
